File: benchmarks/comparable/python/object_tree.py
# object_tree.py - CPython/protoPython twin of benchmarks/comparable/object_tree.scala:
# build a complete binary tree (depth 16, 131071 objects), path-copy its
# leftmost spine, fold both versions. Prints the result on the last line.
# Result: 131071 278364170 725606090.
import sys

sys.setrecursionlimit(10000)
M = 1000000007


class Leaf:
    __slots__ = ("value",)

    def __init__(self, value):
        self.value = value


class Node:
    __slots__ = ("left", "right", "weight")

    def __init__(self, left, right, weight):
        self.left = left
        self.right = right
        self.weight = weight


def build(depth, seed):
    if depth == 0:
        return Leaf(seed % 1000)
    return Node(build(depth - 1, seed * 2 + 1), build(depth - 1, seed * 2 + 2), depth)


def bump(t):
    if isinstance(t, Leaf):
        return Leaf(t.value + 1)
    return Node(bump(t.left), t.right, t.weight)


def count(t):
    if isinstance(t, Leaf):
        return 1
    return 1 + count(t.left) + count(t.right)


def checksum(t):
    if isinstance(t, Leaf):
        return t.value
    return (checksum(t.left) * 31 + checksum(t.right) + t.weight) % M


t = build(16, 0)
t2 = bump(t)
print(count(t), checksum(t), checksum(t2))

File: benchmarks/comparable/object_tree.scala
// EXPECT: 131071 278364170 725606090
// object_tree.scala - a deep immutable object graph: build a complete binary
// tree of case classes (depth 16, 131071 objects), path-copy its leftmost
// spine (the copy shares every right subtree with the original), and fold
// both versions with pattern matching. Twin of
// benchmarks/comparable/python/object_tree.py.

sealed trait Tree
case class Leaf(value: Int) extends Tree
case class Node(left: Tree, right: Tree, weight: Int) extends Tree

def build(depth: Int, seed: Int): Tree =
  if depth == 0 then Leaf(seed % 1000)
  else Node(build(depth - 1, seed * 2 + 1), build(depth - 1, seed * 2 + 2), depth)

def bump(t: Tree): Tree = t match
  case Leaf(v)       => Leaf(v + 1)
  case Node(l, r, w) => Node(bump(l), r, w)

def count(t: Tree): Int = t match
  case Leaf(_)       => 1
  case Node(l, r, _) => 1 + count(l) + count(r)

def checksum(t: Tree): Long = t match
  case Leaf(v)       => v
  case Node(l, r, w) => (checksum(l) * 31 + checksum(r) + w) % 1000000007L

@main def benchObjectTree(): Unit =
  val t = build(16, 0)
  val t2 = bump(t)
  println(count(t).toString + " " + checksum(t) + " " + checksum(t2))
